src/pip/_internal/download.py

```diff
@@ -939,7 +939,10 @@ def _download_http_url(
         type, params = cgi.parse_header(content_disposition)
         # We use ``or`` here because we don't want to use an "empty" value
         # from the filename param.
-        filename = params.get('filename') or filename
+        _filename = params.get('filename')
+        if _filename:
+            _filename = os.path.basename(_filename)
+        filename = _filename or filename
```

Comment:
To increase testability, I would define a `parse_content_disposition(headers, default_filename)` function that returns the file name to use, and then add a couple unit tests of the function using `@pytest.mark.parametrize`.